File: 2 - Sorting and Order Statistics/2.1 - Heap Sort/Code/heapify.py
"""
heapify.py

"""

from .heaps import Heap


class Heapify(Heap):
    """Heapify class"""

    def __init__(self, heap):
        super().__init__(heap)
        # For now, assume all elements in the array are part of the heap
        self.heap_size = len(heap)

    def max_heapify(self, index):
        """Converts a heap that violates the max-heap property into a valid binary heap

        Time complexity: O(log n)
        Space complexity: O(log n)

        Args:
            heap (list[int]): Input binary heap
            index (int): Starting index
        """
        left_child = self.left_child(index)
        right_child = self.right_child(index)

        largest = index

        if left_child < self.heap_size and self.heap[left_child] > self.heap[index]:
            largest = left_child

        if right_child < self.heap_size and self.heap[right_child] > self.heap[largest]:
            largest = right_child

        if largest != index:
            self.heap[index], self.heap[largest] = self.heap[largest], self.heap[index]
            self.max_heapify(largest)

    def min_heapify(self, index):
        """Converts a heap into a minimum binary heap in an analogous manner to that of the
        procedure above"""
        left_child = self.left_child(index)
        right_child = self.right_child(index)

        smallest = index

        if left_child < self.heap_size and self.heap[left_child] < self.heap[index]:
            smallest = left_child

        if (
            right_child < self.heap_size
            and self.heap[right_child] < self.heap[smallest]
        ):
            smallest = right_child

        if smallest != index:
            self.heap[index], self.heap[smallest] = (
                self.heap[smallest],
                self.heap[index],
            )
            self.min_heapify(smallest)

File: 2 - Sorting and Order Statistics/2.1 - Heap Sort/Code/Heapify.mjs
/**
 * Heapify.mjs
 *
 */

import Heap from "./Heaps.mjs";

export default class Heapify extends Heap {
    constructor(heap) {
        super(heap);
        this.heapSize = heap.length;
    }

    maxHeapify(index) {
        let left = this.leftChild(index);
        let right = this.rightChild(index);
        let largest = -1;

        if (left < this.heapSize && this.heap[left] > this.heap[index]) {
            largest = left;
        } else {
            largest = index;
        }

        if (right < this.heapSize && this.heap[right] > this.heap[largest]) {
            largest = right;
        }

        if (largest != index) {
            let temp = this.heap[index];
            this.heap[index] = this.heap[largest];
            this.heap[largest] = temp;
            this.maxHeapify(largest);
        }
    }

    minHeapify(index) {
        let left = this.leftChild(index);
        let right = this.rightChild(index);
        let smallest = index;

        if (left < this.heapSize && this.heap[left] < this.heap[index]) {
            smallest = left;
        } else {
            smallest = index;
        }

        if (right < this.heapSize && this.heap[right] < this.heap[smallest]) {
            smallest = right;
        }

        if (smallest != index) {
            let temp = this.heap[index];
            this.heap[index] = this.heap[smallest];
            this.heap[smallest] = temp;
            this.minHeapify(smallest);
        }
    }
}

const main = () => {
    const maxArray = [16, 4, 10, 14, 7, 9, 3, 2, 8, 1];
    let heapifyMax = new Heapify(maxArray);
    heapifyMax.maxHeapify(1);
    console.log(heapifyMax.heap);

    const minArray = [1, 3, 14, 4, 7, 8, 9, 2, 10, 16];
    let heapifyMin = new Heapify(minArray);
    heapifyMin.minHeapify(2);
    console.log(heapifyMin.heap);
};

// main();
